sharding/xorValidatorShuffler.go

```diff
@@ -0,0 +1,331 @@
+//          a)  In case (shuffled out nodes + new nodes) > (nbShards * perShardHysteresis + minNodesPerShard) then
+//              we need to prepare for a split event, so a higher percentage of nodes need to be directed to the shard
+//              that will be split.
+//          b)  In case (shuffled out nodes + new nodes) < (nbShards * perShardHysteresis) then we need to prepare for a
+//              shard merge event so
+//          c)  No change in the number of shards then nothing extra needs to be done
+func (rxs *randXORShuffler) UpdateNodeLists(args UpdateNodesArgs) (map[uint32][]Validator, map[uint32][]Validator) {
+	var shuffledOutNodes []Validator
+	var newEligible map[uint32][]Validator
+	var newWaiting map[uint32][]Validator
+
+	newNbShards := rxs.computeNewShards(args.eligible, args.waiting, args.newNodes, args.leaving, args.nbShards)
+
+	rxs.mutShufflerParams.RLock()
+	canSplit := rxs.adaptivity && newNbShards > args.nbShards
+	canMerge := rxs.adaptivity && newNbShards < args.nbShards
+	rxs.mutShufflerParams.RUnlock()
+
+	if canSplit {
+		newEligible, newWaiting = rxs.splitShards(args.eligible, args.waiting, newNbShards)
+	}
+	if canMerge {
+		newEligible, newWaiting = rxs.mergeShards(args.eligible, args.waiting, newNbShards)
+	}
+
+	for shard, vList := range newWaiting {
+		nbToRemove := len(vList)
+		if len(args.leaving) < nbToRemove {
+			nbToRemove = len(args.leaving)
+		}
+
+		vList, args.leaving = removeValidatorsFromList(vList, args.leaving, nbToRemove)
+		newWaiting[shard] = vList
+	}
+
+	shuffledOutNodes, newEligible, args.leaving = shuffleOutNodes(args.eligible, args.waiting, args.leaving, args.rand)
+	promoteWaitingToEligible(newEligible, args.waiting)
+	distributeValidators(args.newNodes, newWaiting, args.rand, newNbShards+1)
+	distributeValidators(shuffledOutNodes, newWaiting, args.rand, newNbShards+1)
+
+	return newEligible, newWaiting
+}
+
+// computeNewShards determines the new number of shards based on the number of nodes in the network
+func (rxs *randXORShuffler) computeNewShards(
+	eligible map[uint32][]Validator,
+	waiting map[uint32][]Validator,
+	newNodes []Validator,
+	leavingNodes []Validator,
+	nbShards uint32,
+) uint32 {
+
+	nbEligible := 0
+	nbWaiting := 0
+	for shard := range eligible {
+		nbEligible += len(eligible[shard])
+		nbWaiting += len(waiting[shard])
+	}
+
+	nodesNewEpoch := uint32(nbEligible + nbWaiting + len(newNodes) - len(leavingNodes))
+
+	rxs.mutShufflerParams.RLock()
+	maxNodesMeta := rxs.nodesMeta + rxs.metaHysteresis
+	maxNodesShard := rxs.nodesShard + rxs.shardHysteresis
+	nodesForSplit := (nbShards+1)*maxNodesShard + maxNodesMeta
+	nodesForMerge := nbShards*rxs.nodesShard + rxs.nodesMeta
+	rxs.mutShufflerParams.RUnlock()
+
+	nbShardsNew := nbShards
+	if nodesNewEpoch > nodesForSplit {
+		nbNodesWithoutMaxMeta := nodesNewEpoch - maxNodesMeta
+		nbShardsNew = nbNodesWithoutMaxMeta / maxNodesShard
+
+		return nbShardsNew
+	}
+
+	if nodesNewEpoch == nodesForMerge {
```

Comment:
The merge will never happen in this case, as the nodes for merge will be the min number of nodes / shard multiplied by the number of shards and it can never go below.

I was thinking of adding another hysteresis threshold for merge that will allow some degree of variation.

Reply:
ok